lib/elastic_record/index/bulk_documents.rb

```diff
@@ -0,0 +1,83 @@
+require 'active_support/core_ext/object/to_query'
+
+module ElasticRecord
+  class Index
+    module BulkDocuments
+      def index_document(id, document, doctype: model.doctype, parent: nil, index_name: alias_name)
+        if batch = current_bulk_batch
+          instructions = { _index: index_name, _type: doctype.name, _id: id }
+          instructions[:parent] = parent if parent
+
+          batch << { index: instructions }
+          batch << document
+        else
+          super
```

Comment:
This calls `ElasticRecord::Index::Documents#index_record`?  I see that this supports the current `index.bulk {}` api, but I'm not a huge fan of the overloading/indirection. I would prefer to see these methods split out into a `BulkDocument` instance which is returned when calling `bulk` e.g.

```ruby
index.bulk do |bulk_document|
  bulk_document.index_document '5', color: 'green'
end
```

I find it easier to comprehend and debug a call to a class instance than to a method defined in multiple places.

This would also allow for a threadsafe by default `bulk_stack` that lives on the `BulkDocument` instance instead of `ElasticRecord::Connection`.

Reply:
> This would also allow for a threadsafe by default `bulk_stack` that lives on the `BulkDocument` instance instead of `ElasticRecord::Connection`.

This cannot work, because each model has their own index. The bulk saves must be tracked across all models. For example:
```
bulk do
  save_widget
  save_customer
end
```

The requirement to track the `bulk_document` variable does make sense, but we have code 10 to 20 method calls deep that rely on being inside of a transaction and a bulk index. Introducing the requirement to pass a `bulk_document` variable along through every method call, and requiring a method to even know it is in a bulk write, would introduce a lot of code into the application.

As for the design pattern and indirection, I agree that secondary change of refactoring out the bulk methods is is not a lot better. I was trying to group together code and tests that relate to bulk actions separately from the standard methods. ActiveRecord does this a lot, though the case is often different because it's not in the format of `if bulk new_code else super end`.

There are many examples of this in AR:
https://github.com/rails/rails/blob/master/activerecord/lib/active_record/validations.rb#L45-L53

Reply:
Ah! Thank you for this explanation. Thinking about it's use as similar to an `ActiveRecord::Base#transaction` makes the requirement for this design much more obvious.